// typesync/lib/package-source.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPackageSource = createPackageSource;
const npm_registry_fetch_1 = __importDefault(require("npm-registry-fetch"));
const semver_1 = require("semver");
/**
 * Creates a package source.
 */
function createPackageSource() {
    return {
        /**
         * Fetches info about a package, or `null` if not found.
         */
        fetch: async (name) => {
            const response = await (0, npm_registry_fetch_1.default)(encodeURI(name)).catch((err) => {
                if (err.statusCode === 404) {
                    return null;
                }
                /* istanbul ignore next */
                throw err;
            });
            const data = await response?.json();
            if (!data?.versions) {
                return null;
            }
            const versionIdentifiers = Object.keys(data.versions)
                .sort(semver_1.compare)
                .reverse();
            const versions = versionIdentifiers.map((v) => {
                const item = data.versions[v];
                return {
                    version: item.version,
                    containsInternalTypings: !!item.types || !!item.typings,
                };
            });
            return {
                name: data.name,
                deprecated: Boolean(data.versions[versionIdentifiers[0]].deprecated),
                latestVersion: data['dist-tags'].latest,
                // Sort by version, highest version first.
                versions: versions,
            };
        },
    };
}
