// eslint-plugin-sonarjs/lib/S5256/meta.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.sonarKey = exports.meta = void 0;
// DO NOT EDIT! This file is autogenerated by "npm run generate-meta"
exports.meta = {
    type: 'problem',
    docs: {
        description: 'Tables should have headers',
        recommended: true,
        url: 'https://sonarsource.github.io/rspec/#/rspec/S5256/javascript',
        requiresTypeChecking: false,
    },
};
exports.sonarKey = 'S5256';
