// eslint-plugin-sonarjs/lib/index.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bridgeRules = exports.rules = exports.meta = exports.configs = exports.pluginRules = void 0;
const fs_1 = __importDefault(require("fs"));
const helpers_1 = require("./helpers");
const S2376_1 = require("./S2376"); // accessor-pairs
const S1077_1 = require("./S1077"); // alt-text
const S6827_1 = require("./S6827"); // anchor-has-content
const S6844_1 = require("./S6844"); // anchor-is-valid
const S5850_1 = require("./S5850"); // anchor-precedence
const S3782_1 = require("./S3782"); // argument-type
const S2234_1 = require("./S2234"); // arguments-order
const S3513_1 = require("./S3513"); // arguments-usage
const S3796_1 = require("./S3796"); // array-callback-without-return
const S1528_1 = require("./S1528"); // array-constructor
const S3524_1 = require("./S3524"); // arrow-function-convention
const S2699_1 = require("./S2699"); // assertions-in-tests
const S6333_1 = require("./S6333"); // aws-apigateway-public-api
const S6329_1 = require("./S6329"); // aws-ec2-rds-dms-public
const S6275_1 = require("./S6275"); // aws-ec2-unencrypted-ebs-volume
const S6332_1 = require("./S6332"); // aws-efs-unencrypted
const S6302_1 = require("./S6302"); // aws-iam-all-privileges
const S6304_1 = require("./S6304"); // aws-iam-all-resources-accessible
const S6317_1 = require("./S6317"); // aws-iam-privilege-escalation
const S6270_1 = require("./S6270"); // aws-iam-public-access
const S6308_1 = require("./S6308"); // aws-opensearchservice-domain
const S6303_1 = require("./S6303"); // aws-rds-unencrypted-databases
const S6321_1 = require("./S6321"); // aws-restricted-ip-admin-access
const S6265_1 = require("./S6265"); // aws-s3-bucket-granted-access
const S6249_1 = require("./S6249"); // aws-s3-bucket-insecure-http
const S6281_1 = require("./S6281"); // aws-s3-bucket-public-access
const S6245_1 = require("./S6245"); // aws-s3-bucket-server-encryption
const S6252_1 = require("./S6252"); // aws-s3-bucket-versioning
const S6319_1 = require("./S6319"); // aws-sagemaker-unencrypted-notebook
const S6327_1 = require("./S6327"); // aws-sns-unencrypted-topics
const S6330_1 = require("./S6330"); // aws-sqs-unencrypted-queue
const S1529_1 = require("./S1529"); // bitwise-operators
const S4798_1 = require("./S4798"); // bool-param-default
const S1105_1 = require("./S1105"); // brace-style
const S1472_1 = require("./S1472"); // call-argument-line
const S5742_1 = require("./S5742"); // certificate-transparency
const S6092_1 = require("./S6092"); // chai-determinate-assertion
const S101_1 = require("./S101"); // class-name
const S3525_1 = require("./S3525"); // class-prototype
const S1523_1 = require("./S1523"); // code-eval
const S3776_1 = require("./S3776"); // cognitive-complexity
const S3616_1 = require("./S3616"); // comma-or-logical-or-case
const S124_1 = require("./S124"); // comment-regex
const S6353_1 = require("./S6353"); // concise-regex
const S3973_1 = require("./S3973"); // conditional-indentation
const S5757_1 = require("./S5757"); // confidential-information-logging
const S1848_1 = require("./S1848"); // constructor-for-side-effects
const S5693_1 = require("./S5693"); // content-length
const S5728_1 = require("./S5728"); // content-security-policy
const S3330_1 = require("./S3330"); // cookie-no-httponly
const S2255_1 = require("./S2255"); // cookies
const S5122_1 = require("./S5122"); // cors
const S4502_1 = require("./S4502"); // csrf
const S1541_1 = require("./S1541"); // cyclomatic-complexity
const S3798_1 = require("./S3798"); // declarations-in-global-scope
const S1788_1 = require("./S1788"); // default-param-last
const S1874_1 = require("./S1874"); // deprecation
const S3514_1 = require("./S3514"); // destructuring-assignment-syntax
const S3403_1 = require("./S3403"); // different-types-comparison
const S5247_1 = require("./S5247"); // disabled-auto-escaping
const S5725_1 = require("./S5725"); // disabled-resource-integrity
const S6080_1 = require("./S6080"); // disabled-timeout
const S5743_1 = require("./S5743"); // dns-prefetching
const S5869_1 = require("./S5869"); // duplicates-in-character-class
const S126_1 = require("./S126"); // elseif-without-else
const S5842_1 = require("./S5842"); // empty-string-repetition
const S4787_1 = require("./S4787"); // encryption
const S5542_1 = require("./S5542"); // encryption-secure-mode
const S3723_1 = require("./S3723"); // enforce-trailing-comma
const S6328_1 = require("./S6328"); // existing-groups
const S1067_1 = require("./S1067"); // expression-complexity
const S1451_1 = require("./S1451"); // file-header
const S3317_1 = require("./S3317"); // file-name-differ-from-class
const S2612_1 = require("./S2612"); // file-permissions
const S2598_1 = require("./S2598"); // file-uploads
const S1134_1 = require("./S1134"); // fixme-tag
const S1535_1 = require("./S1535"); // for-in
const S2251_1 = require("./S2251"); // for-loop-increment-sign
const S5732_1 = require("./S5732"); // frame-ancestors
const S1515_1 = require("./S1515"); // function-inside-loop
const S100_1 = require("./S100"); // function-name
const S3800_1 = require("./S3800"); // function-return-type
const S1527_1 = require("./S1527"); // future-reserved-words
const S3531_1 = require("./S3531"); // generator-without-yield
const S4790_1 = require("./S4790"); // hashing
const S5691_1 = require("./S5691"); // hidden-files
const S6754_1 = require("./S6754"); // hook-use-state
const S5254_1 = require("./S5254"); // html-has-lang
const S3785_1 = require("./S3785"); // in-operator-type-error
const S3686_1 = require("./S3686"); // inconsistent-function-call
const S2692_1 = require("./S2692"); // index-of-compare-to-positive-number
const S2092_1 = require("./S2092"); // insecure-cookie
const S5659_1 = require("./S5659"); // insecure-jwt-token
const S3415_1 = require("./S3415"); // inverted-assertion-arguments
const S6477_1 = require("./S6477"); // jsx-key
const S6481_1 = require("./S6481"); // jsx-no-constructed-context-values
const S6749_1 = require("./S6749"); // jsx-no-useless-fragment
const S6853_1 = require("./S6853"); // label-has-associated-control
const S1439_1 = require("./S1439"); // label-position
const S5148_1 = require("./S5148"); // link-with-target-blank
const S1479_1 = require("./S1479"); // max-switch-cases
const S4622_1 = require("./S4622"); // max-union-size
const S4084_1 = require("./S4084"); // media-has-caption
const S1994_1 = require("./S1994"); // misplaced-loop-counter
const S1082_1 = require("./S1082"); // mouse-events-a11y
const S134_1 = require("./S134"); // nested-control-flow
const S2430_1 = require("./S2430"); // new-cap
const S2999_1 = require("./S2999"); // new-operator-misuse
const S4275_1 = require("./S4275"); // no-accessor-field-mismatch
const S3923_1 = require("./S3923"); // no-all-duplicated-branches
const S2871_1 = require("./S2871"); // no-alphabetical-sort
const S6268_1 = require("./S6268"); // no-angular-bypass-sanitization
const S2870_1 = require("./S2870"); // no-array-delete
const S6479_1 = require("./S6479"); // no-array-index-key
const S3579_1 = require("./S3579"); // no-associative-arrays
const S6551_1 = require("./S6551"); // no-base-to-string
const S2424_1 = require("./S2424"); // no-built-in-override
const S1219_1 = require("./S1219"); // no-case-label-in-switch
const S5332_1 = require("./S5332"); // no-clear-text-protocols
const S6079_1 = require("./S6079"); // no-code-after-done
const S1066_1 = require("./S1066"); // no-collapsible-if
const S3981_1 = require("./S3981"); // no-collection-size-mischeck
const S125_1 = require("./S125"); // no-commented-code
const S1854_1 = require("./S1854"); // no-dead-store
const S3001_1 = require("./S3001"); // no-delete-var
const S6957_1 = require("./S6957"); // no-deprecated-react
const S4621_1 = require("./S4621"); // no-duplicate-in-composite
const S1192_1 = require("./S1192"); // no-duplicate-string
const S1871_1 = require("./S1871"); // no-duplicated-branches
const S4143_1 = require("./S4143"); // no-element-overwrite
const S6019_1 = require("./S6019"); // no-empty-after-reluctant
const S6323_1 = require("./S6323"); // no-empty-alternatives
const S4158_1 = require("./S4158"); // no-empty-collection
const S1186_1 = require("./S1186"); // no-empty-function
const S6331_1 = require("./S6331"); // no-empty-group
const S4023_1 = require("./S4023"); // no-empty-interface
const S2187_1 = require("./S2187"); // no-empty-test-file
const S888_1 = require("./S888"); // no-equals-in-for-termination
const S6426_1 = require("./S6426"); // no-exclusive-tests
const S6643_1 = require("./S6643"); // no-extend-native
const S930_1 = require("./S930"); // no-extra-arguments
const S1116_1 = require("./S1116"); // no-extra-semi
const S6788_1 = require("./S6788"); // no-find-dom-node
const S4139_1 = require("./S4139"); // no-for-in-iterable
const S1530_1 = require("./S1530"); // no-function-declaration-in-block
const S2990_1 = require("./S2990"); // no-global-this
const S2137_1 = require("./S2137"); // no-globals-shadowing
const S2589_1 = require("./S2589"); // no-gratuitous-expressions
const S2068_1 = require("./S2068"); // no-hardcoded-credentials
const S1313_1 = require("./S1313"); // no-hardcoded-ip
const S6442_1 = require("./S6442"); // no-hook-setter-in-body
const S1862_1 = require("./S1862"); // no-identical-conditions
const S1764_1 = require("./S1764"); // no-identical-expressions
const S4144_1 = require("./S4144"); // no-identical-functions
const S2486_1 = require("./S2486"); // no-ignored-exceptions
const S2201_1 = require("./S2201"); // no-ignored-return
const S4328_1 = require("./S4328"); // no-implicit-dependencies
const S2703_1 = require("./S2703"); // no-implicit-global
const S4619_1 = require("./S4619"); // no-in-misuse
const S1940_1 = require("./S1940"); // no-inverted-boolean-check
const S2970_1 = require("./S2970"); // no-incomplete-assertions
const S3801_1 = require("./S3801"); // no-inconsistent-returns
const S3402_1 = require("./S3402"); // no-incorrect-string-concat
const S2189_1 = require("./S2189"); // no-infinite-loop
const S5604_1 = require("./S5604"); // no-intrusive-permissions
const S4123_1 = require("./S4123"); // no-invalid-await
const S3516_1 = require("./S3516"); // no-invariant-returns
const S5759_1 = require("./S5759"); // no-ip-forward
const S1119_1 = require("./S1119"); // no-labels
const S6958_1 = require("./S6958"); // no-literal-call
const S6660_1 = require("./S6660"); // no-lonely-if
const S5734_1 = require("./S5734"); // no-mime-sniff
const S4043_1 = require("./S4043"); // no-misleading-array-reverse
const S6544_1 = require("./S6544"); // no-misused-promises
const S5730_1 = require("./S5730"); // no-mixed-content
const S1121_1 = require("./S1121"); // no-nested-assignment
const S3358_1 = require("./S3358"); // no-nested-conditional
const S2004_1 = require("./S2004"); // no-nested-functions
const S1821_1 = require("./S1821"); // no-nested-switch
const S4624_1 = require("./S4624"); // no-nested-template-literals
const S881_1 = require("./S881"); // no-nested-incdec
const S1751_1 = require("./S1751"); // no-one-iteration-loop
const S4036_1 = require("./S4036"); // no-os-command-from-path
const S1226_1 = require("./S1226"); // no-parameter-reassignment
const S1533_1 = require("./S1533"); // no-primitive-wrappers
const S2814_1 = require("./S2814"); // no-redeclare
const S4165_1 = require("./S4165"); // no-redundant-assignments
const S1125_1 = require("./S1125"); // no-redundant-boolean
const S3626_1 = require("./S3626"); // no-redundant-jump
const S4782_1 = require("./S4782"); // no-redundant-optional
const S1110_1 = require("./S1110"); // no-redundant-parentheses
const S6571_1 = require("./S6571"); // no-redundant-type-constituents
const S3827_1 = require("./S3827"); // no-reference-error
const S5736_1 = require("./S5736"); // no-referrer-policy
const S3533_1 = require("./S3533"); // no-require-or-define
const S4324_1 = require("./S4324"); // no-return-type-any
const S5863_1 = require("./S5863"); // no-same-argument-assert
const S3972_1 = require("./S3972"); // no-same-line-conditional
const S6679_1 = require("./S6679"); // no-self-compare
const S1301_1 = require("./S1301"); // no-small-switch
const S105_1 = require("./S105"); // no-tab
const S5257_1 = require("./S5257"); // no-table-as-layout
const S4327_1 = require("./S4327"); // no-this-alias
const S3696_1 = require("./S3696"); // no-throw-literal
const S4822_1 = require("./S4822"); // no-try-promise
const S4623_1 = require("./S4623"); // no-undefined-argument
const S2138_1 = require("./S2138"); // no-undefined-assignment
const S2681_1 = require("./S2681"); // no-unenclosed-multiline-block
const S6486_1 = require("./S6486"); // no-uniq-key
const S6747_1 = require("./S6747"); // no-unknown-property
const S1763_1 = require("./S1763"); // no-unreachable
const S6791_1 = require("./S6791"); // no-unsafe
const S5042_1 = require("./S5042"); // no-unsafe-unzip
const S6478_1 = require("./S6478"); // no-unstable-nested-components
const S3984_1 = require("./S3984"); // no-unthrown-error
const S4030_1 = require("./S4030"); // no-unused-collection
const S905_1 = require("./S905"); // no-unused-expressions
const S1172_1 = require("./S1172"); // no-unused-function-argument
const S1068_1 = require("./S1068"); // no-unused-private-class-members
const S3699_1 = require("./S3699"); // no-use-of-empty-return-value
const S6676_1 = require("./S6676"); // no-useless-call
const S2737_1 = require("./S2737"); // no-useless-catch
const S6647_1 = require("./S6647"); // no-useless-constructor
const S2123_1 = require("./S2123"); // no-useless-increment
const S4335_1 = require("./S4335"); // no-useless-intersection
const S6443_1 = require("./S6443"); // no-useless-react-setstate
const S3504_1 = require("./S3504"); // no-var
const S1526_1 = require("./S1526"); // no-variable-usage-before-declaration
const S6299_1 = require("./S6299"); // no-vue-bypass-sanitization
const S5547_1 = require("./S5547"); // no-weak-cipher
const S4426_1 = require("./S4426"); // no-weak-keys
const S2208_1 = require("./S2208"); // no-wildcard-import
const S2757_1 = require("./S2757"); // non-existent-operator
const S3760_1 = require("./S3760"); // non-number-in-arithmetic-expression
const S2259_1 = require("./S2259"); // null-dereference
const S5264_1 = require("./S5264"); // object-alt-content
const S3498_1 = require("./S3498"); // object-shorthand
const S3757_1 = require("./S3757"); // operation-returning-nan
const S4721_1 = require("./S4721"); // os-command
const S2819_1 = require("./S2819"); // post-message
const S4524_1 = require("./S4524"); // prefer-default-last
const S6572_1 = require("./S6572"); // prefer-enum-initializers
const S4138_1 = require("./S4138"); // prefer-for-of
const S6598_1 = require("./S6598"); // prefer-function-type
const S1488_1 = require("./S1488"); // prefer-immediate-return
const S4156_1 = require("./S4156"); // prefer-namespace-keyword
const S6606_1 = require("./S6606"); // prefer-nullish-coalescing
const S2428_1 = require("./S2428"); // prefer-object-literal
const S6661_1 = require("./S6661"); // prefer-object-spread
const S4634_1 = require("./S4634"); // prefer-promise-shorthand
const S1126_1 = require("./S1126"); // prefer-single-boolean-return
const S6666_1 = require("./S6666"); // prefer-spread
const S6557_1 = require("./S6557"); // prefer-string-starts-ends-with
const S3512_1 = require("./S3512"); // prefer-template
const S4322_1 = require("./S4322"); // prefer-type-guard
const S1264_1 = require("./S1264"); // prefer-while
const S4823_1 = require("./S4823"); // process-argv
const S4507_1 = require("./S4507"); // production-debug
const S2245_1 = require("./S2245"); // pseudo-random
const S1444_1 = require("./S1444"); // public-static-readonly
const S5443_1 = require("./S5443"); // publicly-writable-directories
const S6959_1 = require("./S6959"); // reduce-initial-value
const S6564_1 = require("./S6564"); // redundant-type-aliases
const S5843_1 = require("./S5843"); // regex-complexity
const S4784_1 = require("./S4784"); // regular-expr
const S6440_1 = require("./S6440"); // rules-of-hooks
const S1438_1 = require("./S1438"); // semi
const S5876_1 = require("./S5876"); // session-regeneration
const S3499_1 = require("./S3499"); // shorthand-property-grouping
const S6397_1 = require("./S6397"); // single-char-in-character-classes
const S6035_1 = require("./S6035"); // single-character-alternation
const S5852_1 = require("./S5852"); // slow-regex
const S4818_1 = require("./S4818"); // sockets
const S2392_1 = require("./S2392"); // sonar-block-scoped-var
const S6439_1 = require("./S6439"); // sonar-jsx-no-leaked-render
const S104_1 = require("./S104"); // sonar-max-lines
const S138_1 = require("./S138"); // sonar-max-lines-per-function
const S107_1 = require("./S107"); // sonar-max-params
const S6324_1 = require("./S6324"); // sonar-no-control-regex
const S1534_1 = require("./S1534"); // sonar-no-dupe-keys
const S2639_1 = require("./S2639"); // sonar-no-empty-character-class
const S128_1 = require("./S128"); // sonar-no-fallthrough
const S5856_1 = require("./S5856"); // sonar-no-invalid-regexp
const S109_1 = require("./S109"); // sonar-no-magic-numbers
const S5868_1 = require("./S5868"); // sonar-no-misleading-character-class
const S6326_1 = require("./S6326"); // sonar-no-regex-spaces
const S6441_1 = require("./S6441"); // sonar-no-unused-class-component-methods
const S1481_1 = require("./S1481"); // sonar-no-unused-vars
const S6582_1 = require("./S6582"); // sonar-prefer-optional-chain
const S6759_1 = require("./S6759"); // sonar-prefer-read-only-props
const S6594_1 = require("./S6594"); // sonar-prefer-regexp-exec
const S2077_1 = require("./S2077"); // sql-queries
const S5973_1 = require("./S5973"); // stable-tests
const S4829_1 = require("./S4829"); // standard-input
const S6351_1 = require("./S6351"); // stateful-regex
const S5739_1 = require("./S5739"); // strict-transport-security
const S3003_1 = require("./S3003"); // strings-comparison
const S3854_1 = require("./S3854"); // super-invocation
const S131_1 = require("./S131"); // switch-without-default
const S5256_1 = require("./S5256"); // table-header
const S5260_1 = require("./S5260"); // table-header-reference
const S5958_1 = require("./S5958"); // test-check-exception
const S1135_1 = require("./S1135"); // todo-tag
const S135_1 = require("./S135"); // too-many-break-or-continue-in-loop
const S5867_1 = require("./S5867"); // unicode-aware-regex
const S6535_1 = require("./S6535"); // unnecessary-character-escapes
const S1128_1 = require("./S1128"); // unused-import
const S5860_1 = require("./S5860"); // unused-named-groups
const S4830_1 = require("./S4830"); // unverified-certificate
const S5527_1 = require("./S5527"); // unverified-hostname
const S3500_1 = require("./S3500"); // updated-const-var
const S2310_1 = require("./S2310"); // updated-loop-counter
const S2688_1 = require("./S2688"); // use-isnan
const S4323_1 = require("./S4323"); // use-type-alias
const S1154_1 = require("./S1154"); // useless-string-operation
const S3758_1 = require("./S3758"); // values-not-convertible-to-numbers
const S117_1 = require("./S117"); // variable-name
const S3735_1 = require("./S3735"); // void-use
const S4423_1 = require("./S4423"); // weak-ssl
const S2817_1 = require("./S2817"); // web-sql-database
const S5689_1 = require("./S5689"); // x-powered-by
const S2755_1 = require("./S2755"); // xml-parser-xxe
const S4817_1 = require("./S4817");
const core_1 = require("./core");
const eslint_plugin_jsx_a11y_1 = require("eslint-plugin-jsx-a11y");
const typescript_eslint_1 = require("./typescript-eslint");
const eslint_plugin_import_1 = require("eslint-plugin-import");
const eslint_plugin_react_1 = require("eslint-plugin-react"); // xpath
const bridgeRules = {
    S100: S100_1.rule,
    S101: S101_1.rule,
    S103: core_1.eslintRules['max-len'],
    S104: S104_1.rule,
    S105: S105_1.rule,
    S106: core_1.eslintRules['no-console'],
    S1066: S1066_1.rule,
    S1067: S1067_1.rule,
    S1068: S1068_1.rule,
    S107: S107_1.rule,
    S1077: S1077_1.rule,
    S108: core_1.eslintRules['no-empty'],
    S1082: S1082_1.rule,
    S109: S109_1.rule,
    S1090: eslint_plugin_jsx_a11y_1.rules['iframe-has-title'],
    S1105: S1105_1.rule,
    S1110: S1110_1.rule,
    S1116: S1116_1.rule,
    S1117: typescript_eslint_1.tsEslintRules['no-shadow'],
    S1119: S1119_1.rule,
    S1121: S1121_1.rule,
    S1125: S1125_1.rule,
    S1126: S1126_1.rule,
    S1128: S1128_1.rule,
    S113: core_1.eslintRules['eol-last'],
    S1131: core_1.eslintRules['no-trailing-spaces'],
    S1134: S1134_1.rule,
    S1135: S1135_1.rule,
    S1143: core_1.eslintRules['no-unsafe-finally'],
    S1154: S1154_1.rule,
    S117: S117_1.rule,
    S1172: S1172_1.rule,
    S1186: S1186_1.rule,
    S1192: S1192_1.rule,
    S1199: core_1.eslintRules['no-lone-blocks'],
    S121: core_1.eslintRules['curly'],
    S1219: S1219_1.rule,
    S122: core_1.eslintRules['max-statements-per-line'],
    S1226: S1226_1.rule,
    S124: S124_1.rule,
    S125: S125_1.rule,
    S126: S126_1.rule,
    S1264: S1264_1.rule,
    S128: S128_1.rule,
    S1301: S1301_1.rule,
    S131: S131_1.rule,
    S1313: S1313_1.rule,
    S1314: core_1.eslintRules['no-octal'],
    S1321: core_1.eslintRules['no-with'],
    S134: S134_1.rule,
    S135: S135_1.rule,
    S138: S138_1.rule,
    S139: core_1.eslintRules['line-comment-position'],
    S1438: S1438_1.rule,
    S1439: S1439_1.rule,
    S1440: core_1.eslintRules['eqeqeq'],
    S1441: core_1.eslintRules['quotes'],
    S1442: core_1.eslintRules['no-alert'],
    S1444: S1444_1.rule,
    S1451: S1451_1.rule,
    S1472: S1472_1.rule,
    S1479: S1479_1.rule,
    S1481: S1481_1.rule,
    S1488: S1488_1.rule,
    S1515: S1515_1.rule,
    S1516: core_1.eslintRules['no-multi-str'],
    S1523: S1523_1.rule,
    S1525: core_1.eslintRules['no-debugger'],
    S1526: S1526_1.rule,
    S1527: S1527_1.rule,
    S1528: S1528_1.rule,
    S1529: S1529_1.rule,
    S1530: S1530_1.rule,
    S1533: S1533_1.rule,
    S1534: S1534_1.rule,
    S1535: S1535_1.rule,
    S1536: core_1.eslintRules['no-dupe-args'],
    S1537: typescript_eslint_1.tsEslintRules['comma-dangle'],
    S1539: core_1.eslintRules['strict'],
    S1541: S1541_1.rule,
    S1656: core_1.eslintRules['no-self-assign'],
    S1751: S1751_1.rule,
    S1763: S1763_1.rule,
    S1764: S1764_1.rule,
    S1774: core_1.eslintRules['no-ternary'],
    S1788: S1788_1.rule,
    S1821: S1821_1.rule,
    S1848: S1848_1.rule,
    S1854: S1854_1.rule,
    S1862: S1862_1.rule,
    S1871: S1871_1.rule,
    S1874: S1874_1.rule,
    S1940: S1940_1.rule,
    S1994: S1994_1.rule,
    S2004: S2004_1.rule,
    S2068: S2068_1.rule,
    S2077: S2077_1.rule,
    S2092: S2092_1.rule,
    S2094: typescript_eslint_1.tsEslintRules['no-extraneous-class'],
    S2123: S2123_1.rule,
    S2137: S2137_1.rule,
    S2138: S2138_1.rule,
    S2187: S2187_1.rule,
    S2189: S2189_1.rule,
    S2201: S2201_1.rule,
    S2208: S2208_1.rule,
    S2234: S2234_1.rule,
    S2245: S2245_1.rule,
    S2251: S2251_1.rule,
    S2255: S2255_1.rule,
    S2259: S2259_1.rule,
    S2310: S2310_1.rule,
    S2376: S2376_1.rule,
    S2392: S2392_1.rule,
    S2424: S2424_1.rule,
    S2427: core_1.eslintRules['radix'],
    S2430: S2430_1.rule,
    S2432: core_1.eslintRules['no-setter-return'],
    S2428: S2428_1.rule,
    S2486: S2486_1.rule,
    S2589: S2589_1.rule,
    S2598: S2598_1.rule,
    S2612: S2612_1.rule,
    S2639: S2639_1.rule,
    S2681: S2681_1.rule,
    S2685: core_1.eslintRules['no-caller'],
    S2688: S2688_1.rule,
    S2692: S2692_1.rule,
    S2699: S2699_1.rule,
    S2703: S2703_1.rule,
    S2737: S2737_1.rule,
    S2755: S2755_1.rule,
    S2757: S2757_1.rule,
    S2814: S2814_1.rule,
    S2817: S2817_1.rule,
    S2819: S2819_1.rule,
    S2870: S2870_1.rule,
    S2871: S2871_1.rule,
    S2933: typescript_eslint_1.tsEslintRules['prefer-readonly'],
    S2966: typescript_eslint_1.tsEslintRules['no-non-null-assertion'],
    S2970: S2970_1.rule,
    S2990: S2990_1.rule,
    S2999: S2999_1.rule,
    S3001: S3001_1.rule,
    S3003: S3003_1.rule,
    S3257: typescript_eslint_1.tsEslintRules['no-inferrable-types'],
    S3317: S3317_1.rule,
    S3330: S3330_1.rule,
    S3353: core_1.eslintRules['prefer-const'],
    S3358: S3358_1.rule,
    S3402: S3402_1.rule,
    S3403: S3403_1.rule,
    S3415: S3415_1.rule,
    S3498: S3498_1.rule,
    S3499: S3499_1.rule,
    S3500: S3500_1.rule,
    S3504: S3504_1.rule,
    S3512: S3512_1.rule,
    S3513: S3513_1.rule,
    S3514: S3514_1.rule,
    S3516: S3516_1.rule,
    S3523: core_1.eslintRules['no-new-func'],
    S3524: S3524_1.rule,
    S3525: S3525_1.rule,
    S3531: S3531_1.rule,
    S3533: S3533_1.rule,
    S3579: S3579_1.rule,
    S3616: S3616_1.rule,
    S3626: S3626_1.rule,
    S3686: S3686_1.rule,
    S3696: S3696_1.rule,
    S3699: S3699_1.rule,
    S3758: S3758_1.rule,
    S3723: S3723_1.rule,
    S3735: S3735_1.rule,
    S3757: S3757_1.rule,
    S3760: S3760_1.rule,
    S3776: S3776_1.rule,
    S3782: S3782_1.rule,
    S3785: S3785_1.rule,
    S3786: core_1.eslintRules['no-template-curly-in-string'],
    S3796: S3796_1.rule,
    S3798: S3798_1.rule,
    S3799: core_1.eslintRules['no-empty-pattern'],
    S3800: S3800_1.rule,
    S3801: S3801_1.rule,
    S3812: core_1.eslintRules['no-unsafe-negation'],
    S3827: S3827_1.rule,
    S3834: core_1.eslintRules['no-new-native-nonconstructor'],
    S3854: S3854_1.rule,
    S3863: eslint_plugin_import_1.rules['no-duplicates'],
    S3923: S3923_1.rule,
    S3972: S3972_1.rule,
    S3973: S3973_1.rule,
    S3981: S3981_1.rule,
    S3984: S3984_1.rule,
    S4023: S4023_1.rule,
    S4030: S4030_1.rule,
    S4036: S4036_1.rule,
    S4043: S4043_1.rule,
    S4084: S4084_1.rule,
    S4123: S4123_1.rule,
    S4124: typescript_eslint_1.tsEslintRules['no-misused-new'],
    S4125: core_1.eslintRules['valid-typeof'],
    S4136: typescript_eslint_1.tsEslintRules['adjacent-overload-signatures'],
    S4137: typescript_eslint_1.tsEslintRules['consistent-type-assertions'],
    S4138: S4138_1.rule,
    S4139: S4139_1.rule,
    S4140: core_1.eslintRules['no-sparse-arrays'],
    S4143: S4143_1.rule,
    S4144: S4144_1.rule,
    S4156: S4156_1.rule,
    S4157: typescript_eslint_1.tsEslintRules['no-unnecessary-type-arguments'],
    S4158: S4158_1.rule,
    S4165: S4165_1.rule,
    S4204: typescript_eslint_1.tsEslintRules['no-explicit-any'],
    S4275: S4275_1.rule,
    S4322: S4322_1.rule,
    S4323: S4323_1.rule,
    S4324: S4324_1.rule,
    S4325: typescript_eslint_1.tsEslintRules['no-unnecessary-type-assertion'],
    S4326: core_1.eslintRules['no-return-await'],
    S4327: S4327_1.rule,
    S4328: S4328_1.rule,
    S4335: S4335_1.rule,
    S4423: S4423_1.rule,
    S4426: S4426_1.rule,
    S4502: S4502_1.rule,
    S4507: S4507_1.rule,
    S4524: S4524_1.rule,
    S4619: S4619_1.rule,
    S4621: S4621_1.rule,
    S4622: S4622_1.rule,
    S4623: S4623_1.rule,
    S4624: S4624_1.rule,
    S4634: S4634_1.rule,
    S4721: S4721_1.rule,
    S4782: S4782_1.rule,
    S4784: S4784_1.rule,
    S4787: S4787_1.rule,
    S4790: S4790_1.rule,
    S4798: S4798_1.rule,
    S4817: S4817_1.rule,
    S4818: S4818_1.rule,
    S4822: S4822_1.rule,
    S4823: S4823_1.rule,
    S4829: S4829_1.rule,
    S4830: S4830_1.rule,
    S5042: S5042_1.rule,
    S5122: S5122_1.rule,
    S5148: S5148_1.rule,
    S5247: S5247_1.rule,
    S5254: S5254_1.rule,
    S5256: S5256_1.rule,
    S5257: S5257_1.rule,
    S5260: S5260_1.rule,
    S5264: S5264_1.rule,
    S5332: S5332_1.rule,
    S5443: S5443_1.rule,
    S5527: S5527_1.rule,
    S5542: S5542_1.rule,
    S5547: S5547_1.rule,
    S5604: S5604_1.rule,
    S5659: S5659_1.rule,
    S5689: S5689_1.rule,
    S5691: S5691_1.rule,
    S5693: S5693_1.rule,
    S5725: S5725_1.rule,
    S5728: S5728_1.rule,
    S5730: S5730_1.rule,
    S5732: S5732_1.rule,
    S5734: S5734_1.rule,
    S5736: S5736_1.rule,
    S5739: S5739_1.rule,
    S5742: S5742_1.rule,
    S5743: S5743_1.rule,
    S5757: S5757_1.rule,
    S5759: S5759_1.rule,
    S5842: S5842_1.rule,
    S5843: S5843_1.rule,
    S5850: S5850_1.rule,
    S5852: S5852_1.rule,
    S5856: S5856_1.rule,
    S5860: S5860_1.rule,
    S5863: S5863_1.rule,
    S5867: S5867_1.rule,
    S5868: S5868_1.rule,
    S5869: S5869_1.rule,
    S5876: S5876_1.rule,
    S5958: S5958_1.rule,
    S5973: S5973_1.rule,
    S6019: S6019_1.rule,
    S6035: S6035_1.rule,
    S6079: S6079_1.rule,
    S6080: S6080_1.rule,
    S6092: S6092_1.rule,
    S6245: S6245_1.rule,
    S6249: S6249_1.rule,
    S6252: S6252_1.rule,
    S6265: S6265_1.rule,
    S6268: S6268_1.rule,
    S6270: S6270_1.rule,
    S6275: S6275_1.rule,
    S6281: S6281_1.rule,
    S6299: S6299_1.rule,
    S6302: S6302_1.rule,
    S6303: S6303_1.rule,
    S6304: S6304_1.rule,
    S6308: S6308_1.rule,
    S6317: S6317_1.rule,
    S6319: S6319_1.rule,
    S6321: S6321_1.rule,
    S6323: S6323_1.rule,
    S6324: S6324_1.rule,
    S6325: core_1.eslintRules['prefer-regex-literals'],
    S6326: S6326_1.rule,
    S6327: S6327_1.rule,
    S6328: S6328_1.rule,
    S6329: S6329_1.rule,
    S6330: S6330_1.rule,
    S6331: S6331_1.rule,
    S6332: S6332_1.rule,
    S6333: S6333_1.rule,
    S6351: S6351_1.rule,
    S6353: S6353_1.rule,
    S6397: S6397_1.rule,
    S6426: S6426_1.rule,
    S6435: eslint_plugin_react_1.rules['require-render-return'],
    S6438: eslint_plugin_react_1.rules['jsx-no-comment-textnodes'],
    S6439: S6439_1.rule,
    S6440: S6440_1.rule,
    S6441: S6441_1.rule,
    S6442: S6442_1.rule,
    S6443: S6443_1.rule,
    S6477: S6477_1.rule,
    S6478: S6478_1.rule,
    S6479: S6479_1.rule,
    S6480: eslint_plugin_react_1.rules['jsx-no-bind'],
    S6481: S6481_1.rule,
    S6486: S6486_1.rule,
    S6509: core_1.eslintRules['no-extra-boolean-cast'],
    S6522: core_1.eslintRules['no-import-assign'],
    S6523: core_1.eslintRules['no-unsafe-optional-chaining'],
    S6534: core_1.eslintRules['no-loss-of-precision'],
    S6535: S6535_1.rule,
    S6544: S6544_1.rule,
    S6550: typescript_eslint_1.tsEslintRules['prefer-literal-enum-member'],
    S6551: S6551_1.rule,
    S6557: S6557_1.rule,
    S6564: S6564_1.rule,
    S6565: typescript_eslint_1.tsEslintRules['prefer-return-this-type'],
    S6568: typescript_eslint_1.tsEslintRules['no-confusing-non-null-assertion'],
    S6569: typescript_eslint_1.tsEslintRules['no-unnecessary-type-constraint'],
    S6571: S6571_1.rule,
    S6572: S6572_1.rule,
    S6578: typescript_eslint_1.tsEslintRules['no-duplicate-enum-values'],
    S6582: S6582_1.rule,
    S6583: typescript_eslint_1.tsEslintRules['no-mixed-enums'],
    S6590: typescript_eslint_1.tsEslintRules['prefer-as-const'],
    S6594: S6594_1.rule,
    S6598: S6598_1.rule,
    S6606: S6606_1.rule,
    S6635: core_1.eslintRules['no-constructor-return'],
    S6637: core_1.eslintRules['no-extra-bind'],
    S6638: core_1.eslintRules['no-constant-binary-expression'],
    S6643: S6643_1.rule,
    S6644: core_1.eslintRules['no-unneeded-ternary'],
    S6645: core_1.eslintRules['no-undef-init'],
    S6647: S6647_1.rule,
    S6650: core_1.eslintRules['no-useless-rename'],
    S6653: core_1.eslintRules['prefer-object-has-own'],
    S6654: core_1.eslintRules['no-proto'],
    S6657: core_1.eslintRules['no-octal-escape'],
    S6660: S6660_1.rule,
    S6661: S6661_1.rule,
    S6666: S6666_1.rule,
    S6671: typescript_eslint_1.tsEslintRules['prefer-promise-reject-errors'],
    S6676: S6676_1.rule,
    S6679: S6679_1.rule,
    S6746: eslint_plugin_react_1.rules['no-direct-mutation-state'],
    S6747: S6747_1.rule,
    S6748: eslint_plugin_react_1.rules['no-children-prop'],
    S6749: S6749_1.rule,
    S6750: eslint_plugin_react_1.rules['no-render-return-value'],
    S6754: S6754_1.rule,
    S6756: eslint_plugin_react_1.rules['no-access-state-in-setstate'],
    S6757: eslint_plugin_react_1.rules['no-this-in-sfc'],
    S6759: S6759_1.rule,
    S6761: eslint_plugin_react_1.rules['no-danger-with-children'],
    S6763: eslint_plugin_react_1.rules['no-redundant-should-component-update'],
    S6766: eslint_plugin_react_1.rules['no-unescaped-entities'],
    S6767: eslint_plugin_react_1.rules['no-unused-prop-types'],
    S6770: eslint_plugin_react_1.rules['jsx-pascal-case'],
    S6772: eslint_plugin_react_1.rules['jsx-child-element-spacing'],
    S6774: eslint_plugin_react_1.rules['prop-types'],
    S6775: eslint_plugin_react_1.rules['default-props-match-prop-types'],
    S6788: S6788_1.rule,
    S6789: eslint_plugin_react_1.rules['no-is-mounted'],
    S6790: eslint_plugin_react_1.rules['no-string-refs'],
    S6791: S6791_1.rule,
    S6793: eslint_plugin_jsx_a11y_1.rules['aria-proptypes'],
    S6807: eslint_plugin_jsx_a11y_1.rules['role-has-required-aria-props'],
    S6811: eslint_plugin_jsx_a11y_1.rules['role-supports-aria-props'],
    S6819: eslint_plugin_jsx_a11y_1.rules['prefer-tag-over-role'],
    S6821: eslint_plugin_jsx_a11y_1.rules['aria-role'],
    S6822: eslint_plugin_jsx_a11y_1.rules['no-redundant-roles'],
    S6823: eslint_plugin_jsx_a11y_1.rules['aria-activedescendant-has-tabindex'],
    S6824: eslint_plugin_jsx_a11y_1.rules['aria-unsupported-elements'],
    S6825: eslint_plugin_jsx_a11y_1.rules['no-aria-hidden-on-focusable'],
    S6827: S6827_1.rule,
    S6836: core_1.eslintRules['no-case-declarations'],
    S6840: eslint_plugin_jsx_a11y_1.rules['autocomplete-valid'],
    S6841: eslint_plugin_jsx_a11y_1.rules['tabindex-no-positive'],
    S6842: eslint_plugin_jsx_a11y_1.rules['no-noninteractive-element-to-interactive-role'],
    S6843: eslint_plugin_jsx_a11y_1.rules['no-interactive-element-to-noninteractive-role'],
    S6844: S6844_1.rule,
    S6845: eslint_plugin_jsx_a11y_1.rules['no-noninteractive-tabindex'],
    S6846: eslint_plugin_jsx_a11y_1.rules['no-access-key'],
    S6847: eslint_plugin_jsx_a11y_1.rules['no-noninteractive-element-interactions'],
    S6848: eslint_plugin_jsx_a11y_1.rules['no-static-element-interactions'],
    S6850: eslint_plugin_jsx_a11y_1.rules['heading-has-content'],
    S6851: eslint_plugin_jsx_a11y_1.rules['img-redundant-alt'],
    S6852: eslint_plugin_jsx_a11y_1.rules['interactive-supports-focus'],
    S6853: S6853_1.rule,
    S6859: eslint_plugin_import_1.rules['no-absolute-path'],
    S6861: eslint_plugin_import_1.rules['no-mutable-exports'],
    S6957: S6957_1.rule,
    S6958: S6958_1.rule,
    S6959: S6959_1.rule,
    S878: core_1.eslintRules['no-sequences'],
    S881: S881_1.rule,
    S888: S888_1.rule,
    S905: S905_1.rule,
    S909: core_1.eslintRules['no-continue'],
    S930: S930_1.rule,
};
exports.bridgeRules = bridgeRules;
exports.pluginRules = {};
/**
 * Maps ESLint rule keys to rule implementations in order to preserve the contract of the ESLint plugin
 */
exports.pluginRules['accessor-pairs'] = S2376_1.rule;
exports.pluginRules['alt-text'] = S1077_1.rule;
exports.pluginRules['anchor-has-content'] = S6827_1.rule;
exports.pluginRules['anchor-is-valid'] = S6844_1.rule;
exports.pluginRules['anchor-precedence'] = S5850_1.rule;
exports.pluginRules['argument-type'] = S3782_1.rule;
exports.pluginRules['arguments-order'] = S2234_1.rule;
exports.pluginRules['arguments-usage'] = S3513_1.rule;
exports.pluginRules['array-callback-without-return'] = S3796_1.rule;
exports.pluginRules['array-constructor'] = S1528_1.rule;
exports.pluginRules['arrow-function-convention'] = S3524_1.rule;
exports.pluginRules['assertions-in-tests'] = S2699_1.rule;
exports.pluginRules['aws-apigateway-public-api'] = S6333_1.rule;
exports.pluginRules['aws-ec2-rds-dms-public'] = S6329_1.rule;
exports.pluginRules['aws-ec2-unencrypted-ebs-volume'] = S6275_1.rule;
exports.pluginRules['aws-efs-unencrypted'] = S6332_1.rule;
exports.pluginRules['aws-iam-all-privileges'] = S6302_1.rule;
exports.pluginRules['aws-iam-all-resources-accessible'] = S6304_1.rule;
exports.pluginRules['aws-iam-privilege-escalation'] = S6317_1.rule;
exports.pluginRules['aws-iam-public-access'] = S6270_1.rule;
exports.pluginRules['aws-opensearchservice-domain'] = S6308_1.rule;
exports.pluginRules['aws-rds-unencrypted-databases'] = S6303_1.rule;
exports.pluginRules['aws-restricted-ip-admin-access'] = S6321_1.rule;
exports.pluginRules['aws-s3-bucket-granted-access'] = S6265_1.rule;
exports.pluginRules['aws-s3-bucket-insecure-http'] = S6249_1.rule;
exports.pluginRules['aws-s3-bucket-public-access'] = S6281_1.rule;
exports.pluginRules['aws-s3-bucket-server-encryption'] = S6245_1.rule;
exports.pluginRules['aws-s3-bucket-versioning'] = S6252_1.rule;
exports.pluginRules['aws-sagemaker-unencrypted-notebook'] = S6319_1.rule;
exports.pluginRules['aws-sns-unencrypted-topics'] = S6327_1.rule;
exports.pluginRules['aws-sqs-unencrypted-queue'] = S6330_1.rule;
exports.pluginRules['bitwise-operators'] = S1529_1.rule;
exports.pluginRules['bool-param-default'] = S4798_1.rule;
exports.pluginRules['brace-style'] = S1105_1.rule;
exports.pluginRules['call-argument-line'] = S1472_1.rule;
exports.pluginRules['certificate-transparency'] = S5742_1.rule;
exports.pluginRules['chai-determinate-assertion'] = S6092_1.rule;
exports.pluginRules['class-name'] = S101_1.rule;
exports.pluginRules['class-prototype'] = S3525_1.rule;
exports.pluginRules['code-eval'] = S1523_1.rule;
exports.pluginRules['cognitive-complexity'] = S3776_1.rule;
exports.pluginRules['comma-or-logical-or-case'] = S3616_1.rule;
exports.pluginRules['comment-regex'] = S124_1.rule;
exports.pluginRules['concise-regex'] = S6353_1.rule;
exports.pluginRules['conditional-indentation'] = S3973_1.rule;
exports.pluginRules['confidential-information-logging'] = S5757_1.rule;
exports.pluginRules['constructor-for-side-effects'] = S1848_1.rule;
exports.pluginRules['content-length'] = S5693_1.rule;
exports.pluginRules['content-security-policy'] = S5728_1.rule;
exports.pluginRules['cookie-no-httponly'] = S3330_1.rule;
exports.pluginRules['cookies'] = S2255_1.rule;
exports.pluginRules['cors'] = S5122_1.rule;
exports.pluginRules['csrf'] = S4502_1.rule;
exports.pluginRules['cyclomatic-complexity'] = S1541_1.rule;
exports.pluginRules['declarations-in-global-scope'] = S3798_1.rule;
exports.pluginRules['default-param-last'] = S1788_1.rule;
exports.pluginRules['deprecation'] = S1874_1.rule;
exports.pluginRules['destructuring-assignment-syntax'] = S3514_1.rule;
exports.pluginRules['different-types-comparison'] = S3403_1.rule;
exports.pluginRules['disabled-auto-escaping'] = S5247_1.rule;
exports.pluginRules['disabled-resource-integrity'] = S5725_1.rule;
exports.pluginRules['disabled-timeout'] = S6080_1.rule;
exports.pluginRules['dns-prefetching'] = S5743_1.rule;
exports.pluginRules['duplicates-in-character-class'] = S5869_1.rule;
exports.pluginRules['elseif-without-else'] = S126_1.rule;
exports.pluginRules['empty-string-repetition'] = S5842_1.rule;
exports.pluginRules['encryption'] = S4787_1.rule;
exports.pluginRules['encryption-secure-mode'] = S5542_1.rule;
exports.pluginRules['enforce-trailing-comma'] = S3723_1.rule;
exports.pluginRules['existing-groups'] = S6328_1.rule;
exports.pluginRules['expression-complexity'] = S1067_1.rule;
exports.pluginRules['file-header'] = S1451_1.rule;
exports.pluginRules['file-name-differ-from-class'] = S3317_1.rule;
exports.pluginRules['file-permissions'] = S2612_1.rule;
exports.pluginRules['file-uploads'] = S2598_1.rule;
exports.pluginRules['fixme-tag'] = S1134_1.rule;
exports.pluginRules['for-in'] = S1535_1.rule;
exports.pluginRules['for-loop-increment-sign'] = S2251_1.rule;
exports.pluginRules['frame-ancestors'] = S5732_1.rule;
exports.pluginRules['function-inside-loop'] = S1515_1.rule;
exports.pluginRules['function-name'] = S100_1.rule;
exports.pluginRules['function-return-type'] = S3800_1.rule;
exports.pluginRules['future-reserved-words'] = S1527_1.rule;
exports.pluginRules['generator-without-yield'] = S3531_1.rule;
exports.pluginRules['hashing'] = S4790_1.rule;
exports.pluginRules['hidden-files'] = S5691_1.rule;
exports.pluginRules['hook-use-state'] = S6754_1.rule;
exports.pluginRules['html-has-lang'] = S5254_1.rule;
exports.pluginRules['in-operator-type-error'] = S3785_1.rule;
exports.pluginRules['inconsistent-function-call'] = S3686_1.rule;
exports.pluginRules['index-of-compare-to-positive-number'] = S2692_1.rule;
exports.pluginRules['insecure-cookie'] = S2092_1.rule;
exports.pluginRules['insecure-jwt-token'] = S5659_1.rule;
exports.pluginRules['inverted-assertion-arguments'] = S3415_1.rule;
exports.pluginRules['jsx-key'] = S6477_1.rule;
exports.pluginRules['jsx-no-constructed-context-values'] = S6481_1.rule;
exports.pluginRules['jsx-no-useless-fragment'] = S6749_1.rule;
exports.pluginRules['label-has-associated-control'] = S6853_1.rule;
exports.pluginRules['label-position'] = S1439_1.rule;
exports.pluginRules['link-with-target-blank'] = S5148_1.rule;
exports.pluginRules['max-switch-cases'] = S1479_1.rule;
exports.pluginRules['max-union-size'] = S4622_1.rule;
exports.pluginRules['media-has-caption'] = S4084_1.rule;
exports.pluginRules['misplaced-loop-counter'] = S1994_1.rule;
exports.pluginRules['mouse-events-a11y'] = S1082_1.rule;
exports.pluginRules['nested-control-flow'] = S134_1.rule;
exports.pluginRules['new-cap'] = S2430_1.rule;
exports.pluginRules['new-operator-misuse'] = S2999_1.rule;
exports.pluginRules['no-accessor-field-mismatch'] = S4275_1.rule;
exports.pluginRules['no-all-duplicated-branches'] = S3923_1.rule;
exports.pluginRules['no-alphabetical-sort'] = S2871_1.rule;
exports.pluginRules['no-angular-bypass-sanitization'] = S6268_1.rule;
exports.pluginRules['no-array-delete'] = S2870_1.rule;
exports.pluginRules['no-array-index-key'] = S6479_1.rule;
exports.pluginRules['no-associative-arrays'] = S3579_1.rule;
exports.pluginRules['no-base-to-string'] = S6551_1.rule;
exports.pluginRules['no-built-in-override'] = S2424_1.rule;
exports.pluginRules['no-case-label-in-switch'] = S1219_1.rule;
exports.pluginRules['no-clear-text-protocols'] = S5332_1.rule;
exports.pluginRules['no-code-after-done'] = S6079_1.rule;
exports.pluginRules['no-collapsible-if'] = S1066_1.rule;
exports.pluginRules['no-collection-size-mischeck'] = S3981_1.rule;
exports.pluginRules['no-commented-code'] = S125_1.rule;
exports.pluginRules['no-dead-store'] = S1854_1.rule;
exports.pluginRules['no-delete-var'] = S3001_1.rule;
exports.pluginRules['no-deprecated-react'] = S6957_1.rule;
exports.pluginRules['no-duplicate-in-composite'] = S4621_1.rule;
exports.pluginRules['no-duplicate-string'] = S1192_1.rule;
exports.pluginRules['no-duplicated-branches'] = S1871_1.rule;
exports.pluginRules['no-element-overwrite'] = S4143_1.rule;
exports.pluginRules['no-empty-after-reluctant'] = S6019_1.rule;
exports.pluginRules['no-empty-alternatives'] = S6323_1.rule;
exports.pluginRules['no-empty-collection'] = S4158_1.rule;
exports.pluginRules['no-empty-function'] = S1186_1.rule;
exports.pluginRules['no-empty-group'] = S6331_1.rule;
exports.pluginRules['no-empty-interface'] = S4023_1.rule;
exports.pluginRules['no-empty-test-file'] = S2187_1.rule;
exports.pluginRules['no-equals-in-for-termination'] = S888_1.rule;
exports.pluginRules['no-exclusive-tests'] = S6426_1.rule;
exports.pluginRules['no-extend-native'] = S6643_1.rule;
exports.pluginRules['no-extra-arguments'] = S930_1.rule;
exports.pluginRules['no-extra-semi'] = S1116_1.rule;
exports.pluginRules['no-find-dom-node'] = S6788_1.rule;
exports.pluginRules['no-for-in-iterable'] = S4139_1.rule;
exports.pluginRules['no-function-declaration-in-block'] = S1530_1.rule;
exports.pluginRules['no-global-this'] = S2990_1.rule;
exports.pluginRules['no-globals-shadowing'] = S2137_1.rule;
exports.pluginRules['no-gratuitous-expressions'] = S2589_1.rule;
exports.pluginRules['no-hardcoded-credentials'] = S2068_1.rule;
exports.pluginRules['no-hardcoded-ip'] = S1313_1.rule;
exports.pluginRules['no-hook-setter-in-body'] = S6442_1.rule;
exports.pluginRules['no-identical-conditions'] = S1862_1.rule;
exports.pluginRules['no-identical-expressions'] = S1764_1.rule;
exports.pluginRules['no-identical-functions'] = S4144_1.rule;
exports.pluginRules['no-ignored-exceptions'] = S2486_1.rule;
exports.pluginRules['no-ignored-return'] = S2201_1.rule;
exports.pluginRules['no-implicit-dependencies'] = S4328_1.rule;
exports.pluginRules['no-implicit-global'] = S2703_1.rule;
exports.pluginRules['no-in-misuse'] = S4619_1.rule;
exports.pluginRules['no-incomplete-assertions'] = S2970_1.rule;
exports.pluginRules['no-inconsistent-returns'] = S3801_1.rule;
exports.pluginRules['no-incorrect-string-concat'] = S3402_1.rule;
exports.pluginRules['no-infinite-loop'] = S2189_1.rule;
exports.pluginRules['no-intrusive-permissions'] = S5604_1.rule;
exports.pluginRules['no-invalid-await'] = S4123_1.rule;
exports.pluginRules['no-invariant-returns'] = S3516_1.rule;
exports.pluginRules['no-inverted-boolean-check'] = S1940_1.rule;
exports.pluginRules['no-ip-forward'] = S5759_1.rule;
exports.pluginRules['no-labels'] = S1119_1.rule;
exports.pluginRules['no-literal-call'] = S6958_1.rule;
exports.pluginRules['no-lonely-if'] = S6660_1.rule;
exports.pluginRules['no-mime-sniff'] = S5734_1.rule;
exports.pluginRules['no-misleading-array-reverse'] = S4043_1.rule;
exports.pluginRules['no-misused-promises'] = S6544_1.rule;
exports.pluginRules['no-mixed-content'] = S5730_1.rule;
exports.pluginRules['no-nested-assignment'] = S1121_1.rule;
exports.pluginRules['no-nested-conditional'] = S3358_1.rule;
exports.pluginRules['no-nested-functions'] = S2004_1.rule;
exports.pluginRules['no-nested-incdec'] = S881_1.rule;
exports.pluginRules['no-nested-switch'] = S1821_1.rule;
exports.pluginRules['no-nested-template-literals'] = S4624_1.rule;
exports.pluginRules['no-one-iteration-loop'] = S1751_1.rule;
exports.pluginRules['no-os-command-from-path'] = S4036_1.rule;
exports.pluginRules['no-parameter-reassignment'] = S1226_1.rule;
exports.pluginRules['no-primitive-wrappers'] = S1533_1.rule;
exports.pluginRules['no-redeclare'] = S2814_1.rule;
exports.pluginRules['no-redundant-assignments'] = S4165_1.rule;
exports.pluginRules['no-redundant-boolean'] = S1125_1.rule;
exports.pluginRules['no-redundant-jump'] = S3626_1.rule;
exports.pluginRules['no-redundant-optional'] = S4782_1.rule;
exports.pluginRules['no-redundant-parentheses'] = S1110_1.rule;
exports.pluginRules['no-redundant-type-constituents'] = S6571_1.rule;
exports.pluginRules['no-reference-error'] = S3827_1.rule;
exports.pluginRules['no-referrer-policy'] = S5736_1.rule;
exports.pluginRules['no-require-or-define'] = S3533_1.rule;
exports.pluginRules['no-return-type-any'] = S4324_1.rule;
exports.pluginRules['no-same-argument-assert'] = S5863_1.rule;
exports.pluginRules['no-same-line-conditional'] = S3972_1.rule;
exports.pluginRules['no-self-compare'] = S6679_1.rule;
exports.pluginRules['no-small-switch'] = S1301_1.rule;
exports.pluginRules['no-tab'] = S105_1.rule;
exports.pluginRules['no-table-as-layout'] = S5257_1.rule;
exports.pluginRules['no-this-alias'] = S4327_1.rule;
exports.pluginRules['no-throw-literal'] = S3696_1.rule;
exports.pluginRules['no-try-promise'] = S4822_1.rule;
exports.pluginRules['no-undefined-argument'] = S4623_1.rule;
exports.pluginRules['no-undefined-assignment'] = S2138_1.rule;
exports.pluginRules['no-unenclosed-multiline-block'] = S2681_1.rule;
exports.pluginRules['no-uniq-key'] = S6486_1.rule;
exports.pluginRules['no-unknown-property'] = S6747_1.rule;
exports.pluginRules['no-unreachable'] = S1763_1.rule;
exports.pluginRules['no-unsafe'] = S6791_1.rule;
exports.pluginRules['no-unsafe-unzip'] = S5042_1.rule;
exports.pluginRules['no-unstable-nested-components'] = S6478_1.rule;
exports.pluginRules['no-unthrown-error'] = S3984_1.rule;
exports.pluginRules['no-unused-collection'] = S4030_1.rule;
exports.pluginRules['no-unused-expressions'] = S905_1.rule;
exports.pluginRules['no-unused-function-argument'] = S1172_1.rule;
exports.pluginRules['no-unused-private-class-members'] = S1068_1.rule;
exports.pluginRules['no-use-of-empty-return-value'] = S3699_1.rule;
exports.pluginRules['no-useless-call'] = S6676_1.rule;
exports.pluginRules['no-useless-catch'] = S2737_1.rule;
exports.pluginRules['no-useless-constructor'] = S6647_1.rule;
exports.pluginRules['no-useless-increment'] = S2123_1.rule;
exports.pluginRules['no-useless-intersection'] = S4335_1.rule;
exports.pluginRules['no-useless-react-setstate'] = S6443_1.rule;
exports.pluginRules['no-var'] = S3504_1.rule;
exports.pluginRules['no-variable-usage-before-declaration'] = S1526_1.rule;
exports.pluginRules['no-vue-bypass-sanitization'] = S6299_1.rule;
exports.pluginRules['no-weak-cipher'] = S5547_1.rule;
exports.pluginRules['no-weak-keys'] = S4426_1.rule;
exports.pluginRules['no-wildcard-import'] = S2208_1.rule;
exports.pluginRules['non-existent-operator'] = S2757_1.rule;
exports.pluginRules['non-number-in-arithmetic-expression'] = S3760_1.rule;
exports.pluginRules['null-dereference'] = S2259_1.rule;
exports.pluginRules['object-alt-content'] = S5264_1.rule;
exports.pluginRules['object-shorthand'] = S3498_1.rule;
exports.pluginRules['operation-returning-nan'] = S3757_1.rule;
exports.pluginRules['os-command'] = S4721_1.rule;
exports.pluginRules['post-message'] = S2819_1.rule;
exports.pluginRules['prefer-default-last'] = S4524_1.rule;
exports.pluginRules['prefer-enum-initializers'] = S6572_1.rule;
exports.pluginRules['prefer-for-of'] = S4138_1.rule;
exports.pluginRules['prefer-function-type'] = S6598_1.rule;
exports.pluginRules['prefer-immediate-return'] = S1488_1.rule;
exports.pluginRules['prefer-namespace-keyword'] = S4156_1.rule;
exports.pluginRules['prefer-nullish-coalescing'] = S6606_1.rule;
exports.pluginRules['prefer-object-literal'] = S2428_1.rule;
exports.pluginRules['prefer-object-spread'] = S6661_1.rule;
exports.pluginRules['prefer-promise-shorthand'] = S4634_1.rule;
exports.pluginRules['prefer-single-boolean-return'] = S1126_1.rule;
exports.pluginRules['prefer-spread'] = S6666_1.rule;
exports.pluginRules['prefer-string-starts-ends-with'] = S6557_1.rule;
exports.pluginRules['prefer-template'] = S3512_1.rule;
exports.pluginRules['prefer-type-guard'] = S4322_1.rule;
exports.pluginRules['prefer-while'] = S1264_1.rule;
exports.pluginRules['process-argv'] = S4823_1.rule;
exports.pluginRules['production-debug'] = S4507_1.rule;
exports.pluginRules['pseudo-random'] = S2245_1.rule;
exports.pluginRules['public-static-readonly'] = S1444_1.rule;
exports.pluginRules['publicly-writable-directories'] = S5443_1.rule;
exports.pluginRules['reduce-initial-value'] = S6959_1.rule;
exports.pluginRules['redundant-type-aliases'] = S6564_1.rule;
exports.pluginRules['regex-complexity'] = S5843_1.rule;
exports.pluginRules['regular-expr'] = S4784_1.rule;
exports.pluginRules['pluginRules-of-hooks'] = S6440_1.rule;
exports.pluginRules['semi'] = S1438_1.rule;
exports.pluginRules['session-regeneration'] = S5876_1.rule;
exports.pluginRules['shorthand-property-grouping'] = S3499_1.rule;
exports.pluginRules['single-char-in-character-classes'] = S6397_1.rule;
exports.pluginRules['single-character-alternation'] = S6035_1.rule;
exports.pluginRules['slow-regex'] = S5852_1.rule;
exports.pluginRules['sockets'] = S4818_1.rule;
exports.pluginRules['sonar-block-scoped-var'] = S2392_1.rule;
exports.pluginRules['sonar-jsx-no-leaked-render'] = S6439_1.rule;
exports.pluginRules['sonar-max-lines'] = S104_1.rule;
exports.pluginRules['sonar-max-lines-per-function'] = S138_1.rule;
exports.pluginRules['sonar-max-params'] = S107_1.rule;
exports.pluginRules['sonar-no-control-regex'] = S6324_1.rule;
exports.pluginRules['sonar-no-dupe-keys'] = S1534_1.rule;
exports.pluginRules['sonar-no-empty-character-class'] = S2639_1.rule;
exports.pluginRules['sonar-no-fallthrough'] = S128_1.rule;
exports.pluginRules['sonar-no-invalid-regexp'] = S5856_1.rule;
exports.pluginRules['sonar-no-magic-numbers'] = S109_1.rule;
exports.pluginRules['sonar-no-misleading-character-class'] = S5868_1.rule;
exports.pluginRules['sonar-no-regex-spaces'] = S6326_1.rule;
exports.pluginRules['sonar-no-unused-class-component-methods'] = S6441_1.rule;
exports.pluginRules['sonar-no-unused-vars'] = S1481_1.rule;
exports.pluginRules['sonar-prefer-optional-chain'] = S6582_1.rule;
exports.pluginRules['sonar-prefer-read-only-props'] = S6759_1.rule;
exports.pluginRules['sonar-prefer-regexp-exec'] = S6594_1.rule;
exports.pluginRules['sql-queries'] = S2077_1.rule;
exports.pluginRules['stable-tests'] = S5973_1.rule;
exports.pluginRules['standard-input'] = S4829_1.rule;
exports.pluginRules['stateful-regex'] = S6351_1.rule;
exports.pluginRules['strict-transport-security'] = S5739_1.rule;
exports.pluginRules['strings-comparison'] = S3003_1.rule;
exports.pluginRules['super-invocation'] = S3854_1.rule;
exports.pluginRules['switch-without-default'] = S131_1.rule;
exports.pluginRules['table-header'] = S5256_1.rule;
exports.pluginRules['table-header-reference'] = S5260_1.rule;
exports.pluginRules['test-check-exception'] = S5958_1.rule;
exports.pluginRules['todo-tag'] = S1135_1.rule;
exports.pluginRules['too-many-break-or-continue-in-loop'] = S135_1.rule;
exports.pluginRules['unicode-aware-regex'] = S5867_1.rule;
exports.pluginRules['unnecessary-character-escapes'] = S6535_1.rule;
exports.pluginRules['unused-import'] = S1128_1.rule;
exports.pluginRules['unused-named-groups'] = S5860_1.rule;
exports.pluginRules['unverified-certificate'] = S4830_1.rule;
exports.pluginRules['unverified-hostname'] = S5527_1.rule;
exports.pluginRules['updated-const-var'] = S3500_1.rule;
exports.pluginRules['updated-loop-counter'] = S2310_1.rule;
exports.pluginRules['use-isnan'] = S2688_1.rule;
exports.pluginRules['use-type-alias'] = S4323_1.rule;
exports.pluginRules['useless-string-operation'] = S1154_1.rule;
exports.pluginRules['values-not-convertible-to-numbers'] = S3758_1.rule;
exports.pluginRules['variable-name'] = S117_1.rule;
exports.pluginRules['void-use'] = S3735_1.rule;
exports.pluginRules['weak-ssl'] = S4423_1.rule;
exports.pluginRules['web-sql-database'] = S2817_1.rule;
exports.pluginRules['x-powered-by'] = S5689_1.rule;
exports.pluginRules['xml-parser-xxe'] = S2755_1.rule;
exports.pluginRules['xpath'] = S4817_1.rule;
const rules = {
    ...bridgeRules,
    ...exports.pluginRules,
};
exports.rules = rules;
const recommendedLegacyConfig = { plugins: ['sonarjs'], rules: {} };
const recommendedConfig = {
    name: 'sonarjs/recommended',
    plugins: {
        sonarjs: {
            rules,
        },
    },
    rules: {},
    settings: {
        react: {
            version: '999.999.999',
        },
    },
};
for (const [key, rule] of Object.entries(exports.pluginRules)) {
    const recommended = !!rule.meta.docs?.recommended;
    recommendedConfig.rules[`sonarjs/${key}`] = recommended ? 'error' : 'off';
}
recommendedLegacyConfig.rules = recommendedConfig.rules;
recommendedLegacyConfig.settings = recommendedConfig.settings;
exports.configs = {
    recommended: recommendedConfig,
    'recommended-legacy': recommendedLegacyConfig,
};
/*
 package.json may be in current or parent dir depending on running with ts-jest or built js files
 we need to find it in both cases
 */
const packageJsonPath = (0, helpers_1.findParent)(__dirname, 'package.json');
const { name, version } = (packageJsonPath ? JSON.parse(fs_1.default.readFileSync(packageJsonPath, 'utf-8')) : {});
exports.meta = {
    name,
    version,
};
exports.default = { rules, configs: exports.configs, meta: exports.meta };
__exportStar(require("./helpers"), exports);
